### ISSUE TITLE
Compilation error using `torch_gather` with broadcasted index dimension

### ISSUE THREAD

**Issue by jacobhinkle** [NVIDIA/Fuser#151]
This may be the same underlying issue as #93 but it manifests in a different way. The following example results in an error when compiling kernel:
```c++
// Test unsqueezing the an index vector before gathering, as is done for
// cross_entropy_loss.
TEST_F(NVFuserTest, FusionTorchGatherUnsqueezed_CUDA) {
  int64_t N = 16384;
  int64_t num_classes = 60;
  std::vector<int64_t> input_dims{N, num_classes};
  std::vector<int64_t> index_dims{N};

  at::manual_seed(0);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
  auto options_i = at::TensorOptions().dtype(at::kLong).device(at::kCUDA, 0);

  auto fusion_ptr = std::make_unique<Fusion>();
  Fusion& fusion = *fusion_ptr.get();
  FusionGuard fg(&fusion);

  TensorView* tv_in1 = makeConcreteTensor(input_dims);
  TensorView* tv_idx = makeConcreteTensor(index_dims, DataType::Int);
  fusion.addInput(tv_in1);
  fusion.addInput(tv_idx);

  auto tv_idx_unsqueeze = broadcast(tv_idx, {false, true});
  auto tv_gather = torch_gather(tv_in1, 1, tv_idx_unsqueeze);
  auto tv_gather_squeeze = reshape(tv_gather, {N, 1}, {N});

  fusion.addOutput(tv_gather_squeeze);

  at::Tensor input_1 = at::randn(input_dims, options);
  at::Tensor input_2 = at::randn(index_dims, options);
  at::Tensor input_idx =
      at::randint(0, num_classes, index_dims, options_i);

  auto t_gather = at::gather(input_1, 1, input_idx.unsqueeze(1)).squeeze();

  std::vector<c10::IValue> aten_inputs = {input_1, input_idx};

  FusionExecutorCache executor_cache(std::move(fusion_ptr));
  auto cg_outputs = executor_cache.runFusionWithInputs(aten_inputs);
  testValidate(
      &fusion, cg_outputs, aten_inputs, {t_gather}, __LINE__, __FILE__);
}
```
The generated kernel is
```c++
__global__ void kernel1(Tensor<float, 2> T0, Tensor<int64_t, 1> T1, Tensor<float, 1> T4) {                                               
  NVFUSER_DEFINE_MAGIC_ZERO                                                                                                              
  int i318;                                                                                                                              
  i318 = ((T1.stride[0] * 2) * ((nvfuser_index_t)threadIdx.x)) + ((T1.stride[0] * 256) * ((nvfuser_index_t)blockIdx.x));                 
  int i614;                                                                                                                              
  i614 = ((T0.stride[0] * 2) * ((nvfuser_index_t)threadIdx.x)) + ((T0.stride[0] * 256) * ((nvfuser_index_t)blockIdx.x));                 
  int i802;                                                                                                                              
  i802 = 2 * ((nvfuser_index_t)threadIdx.x);                                                                                             
  int i803;                                                                                                                              
  i803 = 256 * ((nvfuser_index_t)blockIdx.x);                                                                                            
  int i804;                                                                                                                              
  i804 = i802 + i803;                                                                                                                    
  bool b1981;                                                                                                                            
  b1981 = ((1 + i802) + i803) < 16384;                                                                                                   
  if ((((i802 + 1) + i803) < 16384)) {                                                                                                   
    int64_t T5[2];                                                                                                                       
    #pragma unroll                                                                                                                           for(nvfuser_index_t i111 = 0; i111 < 2; ++i111) {                                                                                    
      T5[i111] = 0;                                                                                                                      
    }                                                                                                                                        NVFUSER_UPDATE_MAGIC_ZERO                                                                                                                #pragma unroll                                                                                                                           for(nvfuser_index_t i111 = 0; i111 < 2; ++i111) {                                                                                    
      T5[i111]                                                                                                                                    = T1[(i318 + (T1.stride[0] * (i111 + nvfuser_zero)))];                                                                          
    }                                                                                                                                        NVFUSER_UPDATE_MAGIC_ZERO                                                                                                            
    Array<float, 2, 2> T6;                                                                                                                   #pragma unroll                                                                                                                       
    for(nvfuser_index_t i112 = 0; i112 < 2; ++i112) {                                                                                          float T3[1];                                                                                                                       
      T3[0] = 0;                                                                                                                               T3[0]                                                                                                                              
         = T0[((i614 + (T0.stride[0] * (i112 + nvfuser_zero))) + (T0.stride[1] * T2[0]))];                                                     T6[i112]                                                                                                                           
         = T3[0];                                                                                                                        
    }                                                                                                                                    
    NVFUSER_UPDATE_MAGIC_ZERO                                                                                                            
    loadLocalToGlobal<float, 2, false>( &T4[i804], &T6[0]);                                                                              
  } else {                                                                                                                               
    int64_t T5[2];                                                                                                                       
    #pragma unroll                                                                                                                       
    for(nvfuser_index_t i111 = 0; i111 < 2; ++i111) {                                                                                    
      T5[i111] = 0;                                                                                                                      
    }
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    Array<float, 2, 2> T6;
    #pragma unroll
    for(nvfuser_index_t i112 = 0; i112 < 2; ++i112) {
      float T3[1];
      T3[0] = 0;
      if (b1981) {
        T3[0]
           = T0[((i614 + (T0.stride[0] * (i112 + nvfuser_zero))) + (T0.stride[1] * T2[0]))];
      }
      T6[i112]
         = T3[0];
    }
    NVFUSER_UPDATE_MAGIC_ZERO
    if (b1981) {                                     
      loadLocalToGlobal<float, 2, false>( &T4[i804], &T6[0]);                                                                            
    }                                                
  }                                                  
}                                                    
}                                                    
  /*           
CUDA NVRTC compile error: __tmp_kernel1.cu(9404): error: identifier "T2" is undefined                     
           = T0[((i614 + (T0.stride[0] * (i112 + nvfuser_zero))) + (T0.stride[1] * T2[0]))];              
                                                                                   ^                      
                                                     
__tmp_kernel1.cu(9386): warning #550-D: variable "T5" was set but never used                              
      int64_t T5[2];                                 
              ^                                      
                                                     
Remark: The warnings can be suppressed with "-diag-suppress <warning-number>"
*/
```
Removing the reshape does not change the error.

**Comment by jacobhinkle** [comment]
In the above, we should be replacing `T2_l[0] view(T2)` with `T5[i111] view(T5)`. I think the issue is that this comes inside a `TensorIndex` representing `T0_g`, and `replaceInputsInExpr` is not processing the definitions of `TensorIndex`'s `index_`.  The relevant code is in `FuseBroadcastWithWarpReduce` which is a `kir::IrVisitor`. I may be mistaken @naoyam, but is the problem that `IrVisitor` does not visit `TensorIndex` in this case, or that we do not explicitly handle `TensorIndex` inside `FuseBroadcastWithWarpReduce` in order to specify that it should be replaced?

**Comment by naoyam** [comment]
Seems like it. We do have a "TODO":

https://github.com/NVIDIA/Fuser/blob/main/csrc/lower2device.cpp#L487-L489

Where's your bug-fix PR??? 😄

**Comment by naoyam** [comment]
On a related note, these lowering passes needs more structured organization rather than just we putting them in a straight line function. Each pass has its own dependencies, so much like `IterVisitor`, there should be a pass manager that automatically resolves dependencies and order them in a deterministic and correct way. We started this with just a handful of passes, which was trivial to manage, but not anymore.

**Comment by jacobhinkle** [comment]
> Seems like it. We do have a "TODO":

Haha indeed! 

> Where's your bug-fix PR??? smile
 
:-D Workin' on it

**Comment by jacobhinkle** [comment]
> a pass manager that automatically resolves dependencies 

Oh that's interesting. I could have a look at that. We would define and register classes that provide a pass, with a partial ordering given by their declared dependencies on other passes.

**Comment by jjsjann123** [comment]
Lol I'm running into this as well for my crossEntropyLoss use case.

repro script:
```
import torch
from nvfuser import FusionDefinition, DataType


inputs = [
    torch.randn(4, device='cuda'),
    torch.randn(3, 4, 4, device='cuda'),
    torch.randint(0, 4, (3, 4), device="cuda"),
]

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguous=[True, True, True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguous=[True, True], dtype=DataType.Int, is_cpu=False)
    T3 = fd.ops.reshape(T0, original_shape=[4], new_shape=[4, 1])
    T4 = fd.ops.max(T1, axes=[1], keepdim=False, dtype=DataType.Null)
    T5 = fd.ops.broadcast_in_dim(T4, output_shape=[3, 1, 4], broadcast_dims=[0, 2])
    T6 = fd.ops.broadcast_in_dim(T5, output_shape=[3, 4, 4], broadcast_dims=[0, 1, 2])
    T7 = fd.ops.sub(T1, T6)
    T8 = fd.ops.exp(T7)
    T9 = fd.ops.sum(T8, axes=[1], keepdim=False, dtype=DataType.Null)
    T10 = fd.ops.broadcast_in_dim(T9, output_shape=[3, 1, 4], broadcast_dims=[0, 2])
    T11 = fd.ops.broadcast_in_dim(T10, output_shape=[3, 4, 4], broadcast_dims=[0, 1, 2])
    T12 = fd.ops.div(T8, T11)
    T13 = fd.ops.log(T12)
    T14 = fd.ops.neg(T13)
    T15 = fd.ops.broadcast_in_dim(T3, output_shape=[3, 4, 4], broadcast_dims=[1, 2])
    T16 = fd.ops.mul(T14, T15)
    T17 = fd.ops.reshape(T2, original_shape=[3, 4], new_shape=[3, 1, 4])
    T18 = fd.ops.gather(T16, T17, dim=1)
    fd.add_output(T18)

f = FusionDefinition()

print("tensor 0")
print(inputs[0].shape)
print(inputs[0].stride())

with FusionDefinition() as fd:
    nvfuser_fusion_id0(fd)

print(fd)
for _ in range(5) :
    o = fd.execute(inputs)

print(fd)
print(o[0])
```

Strangely if I enable ignore_index, codegen seems to just work... :exploding_head: 

here's the fd with the mask for ignore_index
```
root@ef02a1539a21:/jiej/playground# PYTORCH_NVFUSER_DUMP=python_definition python thunder_crossEntropyLoss.py
input shape:  torch.Size([3, 4, 4])
target shape:  torch.Size([3, 4])
weight shape:  torch.Size([4])
output shape:  torch.Size([3, 4])

def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(symbolic_sizes=[-1], contiguous=[True], dtype=DataType.Float, is_cpu=False)
    T1 = fd.define_tensor(symbolic_sizes=[-1, -1, -1], contiguous=[True, True, True], dtype=DataType.Float, is_cpu=False)
    T2 = fd.define_tensor(symbolic_sizes=[-1, -1], contiguous=[True, True], dtype=DataType.Int, is_cpu=False)
    T3 = fd.ops.reshape(T0, original_shape=[4], new_shape=[4, 1])
    T4 = fd.ops.max(T1, axes=[1], keepdim=False, dtype=DataType.Null)
    T5 = fd.ops.broadcast_in_dim(T4, output_shape=[3, 1, 4], broadcast_dims=[0, 2])
    T6 = fd.ops.broadcast_in_dim(T5, output_shape=[3, 4, 4], broadcast_dims=[0, 1, 2])
    T7 = fd.ops.sub(T1, T6)
    T8 = fd.ops.exp(T7)
    T9 = fd.ops.sum(T8, axes=[1], keepdim=False, dtype=DataType.Null)
    T10 = fd.ops.broadcast_in_dim(T9, output_shape=[3, 1, 4], broadcast_dims=[0, 2])
    T11 = fd.ops.broadcast_in_dim(T10, output_shape=[3, 4, 4], broadcast_dims=[0, 1, 2])
    T12 = fd.ops.div(T8, T11)
    T13 = fd.ops.log(T12)
    T14 = fd.ops.neg(T13)
    T15 = fd.ops.broadcast_in_dim(T3, output_shape=[3, 4, 4], broadcast_dims=[1, 2])
    T16 = fd.ops.mul(T14, T15)
    T17 = fd.ops.reshape(T2, original_shape=[3, 4], new_shape=[3, 1, 4])
    T18 = fd.ops.gather(T16, T17, dim=1)
    S19 = fd.define_constant(3, dtype=DataType.Int)
    T20 = fd.ops.eq(T17, S19)
    S21 = fd.define_constant(0.00000, dtype=DataType.Double)
    T22 = fd.ops.where(T20, S21, T18)
    fd.add_output(T22)
```

**Comment by jacobhinkle** [comment]
I'm working on moving the fuseWarpReduce lowering pass earlier and combining it with the grid reduce graph rewriting pass to avoid the missing allocation. Note in my original problem just adding the index tensor as an output (I believe; have to check again) causes the code to execute properly too. Ignore_index might just be changing the uses() of the reduction tensor or something that causes that fusion pass the miss, or maybe the sizes are such that it's no longer a single warp reduction. In any case thanks, I'll look again in the AM.
